cpp/inc/bond/protocol/simple_json_reader.h

```diff
@@ -179,15 +195,190 @@ class SimpleJsonReader
+            }
+        }
+
+        StreamHolder& operator=(const StreamHolder& that)
+        {
+            _parent = that._parent;
+
+            if (that.IsParent())
+            {
+                _input.emplace(that._input.value());
+                _stream.emplace(that._stream.value(), _input.value());
+            }
+            else
+            {
+                _input = boost::none;
+                _stream = boost::none;
+            }
+
+            BOOST_ASSERT(IsParent() == that.IsParent());
+            BOOST_ASSERT(IsChild() == that.IsChild());
+
+            return *this;
+        }
+
+        StreamHolder& operator=(StreamHolder&& that)
+        {
+            _parent = that._parent;
+
+            if (that.IsParent())
+            {
+                _input = std::move(that._input);
+                _stream.emplace(that._stream.value(), _input.value());
+
+                BOOST_ASSERT(IsParent());
+                BOOST_ASSERT(!IsChild());
+            }
+            else
+            {
+                _input = boost::none;
+                _stream = boost::none;
+
+                BOOST_ASSERT(!IsParent());
+                BOOST_ASSERT(IsChild());
+            }
+
+            return *this;
+        }
+
+        bool IsParent() const
+        {
+            return static_cast<bool>(_input) && static_cast<bool>(_stream);
+        }
+
+        bool IsChild() const
+        {
+            return _parent != nullptr;
+        }
+
+        StreamHolder* GetParent()
+        {
+            if (IsParent())
+            {
+                return this;
+            }
+
+            BOOST_ASSERT(IsChild());
+            return _parent;
+        }
+
+        const Buffer& GetBuffer() const
+        {
+            if (IsParent())
+            {
+                return _input.value();
+            }
+            else
+            {
+                return _parent->GetBuffer();
+            }
+        }
+
+        Buffer& GetBuffer()
+        {
+            if (IsParent())
+            {
+                return _input.value();
+            }
+            else
+            {
+                return _parent->GetBuffer();
+            }
+        }
+
+        const detail::RapidJsonInputStream<Buffer>& GetStream() const
+        {
+            if (IsParent())
+            {
+                return _stream.value();
+            }
+            else
+            {
+                return _parent->GetStream();
+            }
+        }
+
+        detail::RapidJsonInputStream<Buffer>& GetStream()
+        {
+            if (IsParent())
+            {
+                return _stream.value();
+            }
+            else
+            {
+                return _parent->GetStream();
+            }
+        }
+
+    private:
+        // NOTE: the ctors rely on _input being initialized before _stream
+        boost::optional<Buffer> _input;
+        boost::optional<detail::RapidJsonInputStream<Buffer>> _stream;
```

Comment:
I think you should keep a single `optional` member. Also, maybe changing the `detail::RapidJsonInputStream` by making it own the `Buffer` instead of referring to it may simplify. I don't see it ever referring anything else but the neighboring `_input`.

Reply:
Hmmm... Let me see.